### ISSUE TITLE
Cannot use latest.firefox-beta-bin, list index 0 is out of bounds

### ISSUE THREAD

**Issue by cole-h** [mozilla/nixpkgs-mozilla#231]
If I try to use `latest.firefox-beta-bin`, I get a `list index 0 is out of bounds` error. Full trace:

```
vin@scadrial ~> nix-shell -p latest.firefox-beta-bin --show-trace
error: while evaluating the attribute 'buildInputs' of the derivation 'shell' at /nix/store/l6ajjfl8flrfi230nvqd5xgsqnwafx43-nixos-20.09pre229870.22c98819ccd/nixos/pkgs/build-support/trivial-builders.nix:7:7:
while evaluating the attribute 'buildCommand' of the derivation 'firefox-bin-79.0b6' at /nix/store/l6ajjfl8flrfi230nvqd5xgsqnwafx43-nixos-20.09pre229870.22c98819ccd/nixos/pkgs/stdenv/generic/make-derivation.nix:192:11:
while evaluating the attribute 'src' of the derivation 'firefox-release-bin-unwrapped-79.0b6' at /nix/store/l6ajjfl8flrfi230nvqd5xgsqnwafx43-nixos-20.09pre229870.22c98819ccd/nixos/pkgs/applications/networking/browsers/firefox-bin/default.nix:88:10:
while evaluating 'fetchVersion' at /nix/store/vsipyf7h2f9av7m4mvbaaqj6dmh9h8zi-source/firefox-overlay.nix:106:18, called from /nix/store/vsipyf7h2f9av7m4mvbaaqj6dmh9h8zi-source/firefox-overlay.nix:148:13:
while evaluating anonymous function at /nix/store/l6ajjfl8flrfi230nvqd5xgsqnwafx43-nixos-20.09pre229870.22c98819ccd/nixos/pkgs/build-support/fetchurl/default.nix:38:1, called from /nix/store/vsipyf7h2f9av7m4mvbaaqj6dmh9h8zi-source/firefox-overlay.nix:108:7:
while evaluating the attribute 'sha512' at /nix/store/vsipyf7h2f9av7m4mvbaaqj6dmh9h8zi-source/firefox-overlay.nix:60:9:
while evaluating 'sha512Of' at /nix/store/vsipyf7h2f9av7m4mvbaaqj6dmh9h8zi-source/firefox-overlay.nix:55:28, called from /nix/store/vsipyf7h2f9av7m4mvbaaqj6dmh9h8zi-source/firefox-overlay.nix:60:18:
while evaluating 'extractSha512Sum' at /nix/store/vsipyf7h2f9av7m4mvbaaqj6dmh9h8zi-source/firefox-overlay.nix:33:34, called from /nix/store/vsipyf7h2f9av7m4mvbaaqj6dmh9h8zi-source/firefox-overlay.nix:55:34:
list index 0 is out of bounds, at /nix/store/vsipyf7h2f9av7m4mvbaaqj6dmh9h8zi-source/firefox-overlay.nix:39:24
```

This has been happening for at least a week (I was holding out for it to magically fix itself, but that hasn't happened yet).

**Comment by duog** [comment]
It seems the format of SHA512SUMS
between
https://download.cdn.mozilla.net/pub/firefox/releases/78.0b9/SHA512SUMS
and
https://download.cdn.mozilla.net/pub/firefox/releases/79.0b1/SHA512SUMS

**Comment by cole-h** [comment]
Thanks for figuring that out. With the following patch applied locally, I can now use Firefox Beta again.

```diff
diff --git a/firefox-overlay.nix b/firefox-overlay.nix
index f8b62fc..985f5e2 100644
--- a/firefox-overlay.nix
+++ b/firefox-overlay.nix
@@ -36,9 +36,9 @@ let
     # Nix 2.0 have an bug in `builtins.match` (see https://github.com/NixOS/nix/issues/2147).
     # So I made separate logic for Nix 1.x and Nix 2.0.
     if builtins ? split then
-      substring 0 128 (head
+      substring 2 128 (head
         (super.lib.filter
-          (s: isString s && substring 128 (stringLength s) s == "  ${file}")
+          (s: isString s && substring 131 (stringLength s) s == "  ${file}")
           (split "\n" sha512sums)))
     else
       head (match ".*[\n]([0-9a-f]*)  ${file}.*" sha512sums);
```

**Comment by duog** [comment]
This will break betas  < 79

**Comment by cole-h** [comment]
Right, which is why I haven't PR'd it. Would be nice if Mozilla could fix their output format... Unfortunately, I don't know where that's generated (nor do I want to learn Mercurial), so I won't pursue that angle.

**Comment by duog** [comment]
I've filed a bug https://bugzilla.mozilla.org/show_bug.cgi?id=1653190

**Comment by garbas** [comment]
I would assume switch to python 3 caused this.

I think some of the string writing need to use "six.ensure_str" when dealing with strings somewhere around https://searchfox.org/mozilla-central/source/testing/mozharness/scripts/release/generate-checksums.py

Maybe @rail can give a hand.

**Comment by rail** [comment]
Eeew, another python 3 issue. :) Let me handle this.

**Comment by rail** [comment]
> I've filed a bug https://bugzilla.mozilla.org/show_bug.cgi?id=1653190

Thanks for filing!

**Comment by rail** [comment]
It should be fixed now, https://download.cdn.mozilla.net/pub/firefox/releases/80.0b1/SHA512SUMS looks better. Can someone verify this?

**Comment by cole-h** [comment]
Definitely fixed -- posting from a newly updated Firefox 80.0b1! Thanks!

**Comment by rail** [comment]
\o/

**Comment by danieldk** [comment]
Unfortunately,

https://download.cdn.mozilla.net/pub/firefox/releases/79.0/SHA512SUMS

is also broken. So, installing the latest stable from the overlay gives the same error.

**Comment by rail** [comment]
IIRC, it was too late for the release build to get the fix in. It should be fixed in either in 80.0 or 79.0.1.
